File: pre/c基础/code/GifFrame.java
package com.zero.gifframe;

import android.graphics.Bitmap;
import android.graphics.PixelFormat;
import android.os.FileObserver;
import android.support.annotation.Nullable;

import java.io.InputStream;


public class GifFrame {


    static {
        System.loadLibrary("native-lib");
    }

    private final long nativeHandle;
    private final int width;
    private final int height;
    private final int frameCount;

    public static GifFrame decodeStream(InputStream stream) {
        byte[] buffer = new byte[16 * 1024];
        return nativeDecodeStream(stream, buffer);
    }

    private static native GifFrame nativeDecodeStream(InputStream stream, byte[] buffer);

    private native long nativeGetFrame(long nativeHandle, Bitmap bitmap, int frameIndex);

    private GifFrame(long nativeHandle, int width, int height, int frameCount) {
        this.nativeHandle = nativeHandle;
        this.width = width;
        this.height = height;
        this.frameCount = frameCount;

    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public int getFrameCount() {
        return frameCount;
    }

    public long getFrame(Bitmap bitmap, int frameIndex) {
        return nativeGetFrame(nativeHandle, bitmap, frameIndex);
    }

}

File: source/Giflib1/app/src/main/java/com/zero/giflib1/gif/GifFrame.kt
package com.zero.giflib1.gif

import android.content.Context
import android.content.res.AssetManager
import android.graphics.Bitmap
import android.util.Log
import java.io.InputStream

class GifFrame //JNI 创建java对象
private constructor(
    //C++ 一帧的对象 指针
    var nativeHandle: Long,
    val width: Int,
    val height: Int,
    val frameCount: Int
) {
    constructor() : this(0, 0, 0, 0)


    companion object {
        init {
            System.loadLibrary("native-lib")
        }

        const val TAG = "Zero"

        @JvmStatic
        private external fun nativeDecodeStream(
            stream: InputStream,
            buffer: ByteArray
        ): GifFrame

        @JvmStatic
        private external fun nativeDecodeStreamJNI(
            assetManager: AssetManager?,
            gifPath: String
        ): GifFrame

        @JvmStatic
        private external fun nativeGetFrame(
            nativeHandle: Long,
            bitmap: Bitmap,
            frameIndex: Int
        ): Long

        fun decodeStream(stream: InputStream): GifFrame {
            val buffer = ByteArray(16 * 1024)
            return nativeDecodeStream(stream, buffer)
        }

        fun decodeStream(context: Context?, fileName: String): GifFrame {
            return nativeDecodeStreamJNI(
                context?.assets,
                fileName
            )
        }
    }

    fun getFrame(bitmap: Bitmap, frameIndex: Int): Long {
        Log.i(TAG,"getFrame")
        return nativeGetFrame(nativeHandle, bitmap, frameIndex)
    }

}